accepted/0000-cve-auditing-with-oval.md

```diff
@@ -0,0 +1,366 @@
+sources. For example, instead of downloading OVAL data of SUSE from the official provider
+at: https://ftp.suse.com/pub/projects/security/oval, users could host their own server that acts as a proxy between the
+official repository and their Uyuni server. Configurations will be deployed on the Uyuni server in a user-friendly
+format to allow users to edit it.
+
+#### OVAL Synchronization
+
+Maintaining up-to-date OVAL data is essential for accurate CVE audits, that's why we need to periodically synchronize
+them. Depending on how many distinct operating systems users' systems use and the size of OVAL files, synchronization
+could take some time. Running the synchronization **once a day** seems like a good default since we can't predict the
+scale users will operate on.
+
+In the best case, all of the systems will be using the same operating system .e.g. openSUSE Leap 15.4, in that case
+synchronization will take only a few seconds. And in the worst case, systems' will be using 22 different operating
+systems and thus synchronization could take up to ~30 minutes.
+
+Needless to say we won't be synchronizing OVALs for all supported operating systems; we only need to synchronize the
+operating systems used by registered clients.
+
+In certain scenarios, syncing OVALs may be redundant if the OVAL didn't change since last sync. Some OVAL repositories
+include a manifest file containing hashes of all OVAL files. By comparing these hashes with the synced OVAL hashes,
+synchronization can be selectively performed only when hashes are different. In cases where a manifest file is absent,
+we can scarp the HTML of the OVAL repository server and look for any clues to decide whether we need to resync or not.
+
+### OVAL DAO
+
+The OVAL DAO provides a unified interface of all the operations that can be applied on the OVAL data. It is the only way
+to query or modify the OVAL data from the database.
+
+As a starter, the DAO will provide the following methods:
+
+```java
+public class OVALCachingFactory extends HibernateFactory {
+    public static void savePlatformsVulnerablePackages(OvalRootType rootType) {
+    }
+
+    public static List<VulnerablePackage> getVulnerablePackagesByProductAndCve(String productCpe, String cve) {
+    }
+
+    public static boolean checkOVALAvailability(String cpe) {
+    }
+}
+```
+
+### Database Storage of OVAL
+
+Once the OVAL file is parsed, the next step is to store it in a database. A relational database, such as PostgreSQL,
+seems like an obvious choice given that it is the default database used by Uyuni.
+
+#### Database Schema
+
+The database employs a straightforward schema consisting of three new tables. Notably, there is no need to create a CVE
+table, as we will use the existing `rhnCVE` table integrated into Uyuni. Here's the purpose of each of these tables:
+
+1. **suseOVALVulnerablePackage**: This table represents a single vulnerable package along with its corresponding fix
```

Comment:
Can be fixversion be null, for example for CVE that where detected but no fix was yet released?

Reply:
Yes, I remember trying to use a non-nullable string and set it to an empty string when fix version is unavailable but there was a database check that prevented it.

Reply:
ok, thank sounds good. I think we should have it as null if no fix version is available.